pinot-broker/src/main/java/org/apache/pinot/broker/routing/segmentpruner/TimeRangeSegmentPruner.java

```diff
@@ -0,0 +1,348 @@
+
+
+/**
+ * The {@code TimeRangeSegmentPruner} prunes segments based on their time column start & end time metadata stored in ZK. The pruner
+ * supports queries with filter (or nested filter) of EQUALITY and RANGE predicates.
+ */
+public class TimeRangeSegmentPruner implements SegmentPruner {
+  private static final Logger LOGGER = LoggerFactory.getLogger(TimeRangeSegmentPruner.class);
+  private static final long MAX_END_TIME = Long.MAX_VALUE;
+  private static final long MIN_START_TIME = 0;
+
+  private final String _tableNameWithType;
+  private final ZkHelixPropertyStore _propertyStore;
+  private final String _segmentZKMetadataPathPrefix;
+  private final String _timeColumn;
+  private final TimeUnit _timeUnit;
+
+  private volatile IntervalST _timeRangeMSToSegmentSearchTree;
+
+  public TimeRangeSegmentPruner(TableConfig tableConfig, ZkHelixPropertyStore<ZNRecord> propertyStore) {
+    _tableNameWithType = tableConfig.getTableName();
+    _propertyStore = propertyStore;
+    _segmentZKMetadataPathPrefix = ZKMetadataProvider.constructPropertyStorePathForResource(_tableNameWithType) + "/";
+    _timeColumn = tableConfig.getValidationConfig().getTimeColumnName();
+    Preconditions
+        .checkNotNull(_timeColumn, "Time column must be configured in table config for table: %s", _tableNameWithType);
+
+    Schema schema = ZKMetadataProvider.getTableSchema(_propertyStore, _tableNameWithType);
+    Preconditions.checkState(schema != null, "Failed to find schema for table: %s", _tableNameWithType);
+    DateTimeFieldSpec dateTimeSpec = schema.getSpecForTimeColumn(_timeColumn);
+    Preconditions.checkNotNull(dateTimeSpec, "Field spec must be specified in schema for time column: %s of table: %s",
+        _timeColumn, _tableNameWithType);
+    DateTimeFormatSpec formatSpec = new DateTimeFormatSpec(dateTimeSpec.getFormat());
+    _timeUnit = formatSpec.getColumnUnit();
+    Preconditions
+        .checkNotNull(_timeUnit, "Time unit must be configured in the field spec for time column: %s of table: %s",
+            _timeColumn, _tableNameWithType);
+  }
+
+  @Override
+  public void init(ExternalView externalView, IdealState idealState, Set<String> onlineSegments) {
+    onExternalViewChange(externalView, idealState, onlineSegments);
+  }
+
+  @Override
+  public void onExternalViewChange(ExternalView externalView, IdealState idealState, Set<String> onlineSegments) {
+    updateTimeRangeMSToSegmentSearchTree(onlineSegments);
+  }
+
+  private void updateTimeRangeMSToSegmentSearchTree(Set<String> onlineSegments) {
+    List<String> segments = new ArrayList<>(onlineSegments);
+    // atomic swap _timeRangeToSegmentMap for input online segments
+    List<String> segmentZKMetadataPaths = new ArrayList<>();
+    for (String segment: segments) {
+      segmentZKMetadataPaths.add(_segmentZKMetadataPathPrefix + segment);
+    }
+
+    List<ZNRecord> znRecords = _propertyStore.get(segmentZKMetadataPaths, null, AccessOption.PERSISTENT, true);
```

Comment:
We don't want to read the ZK metadata for all segments for each callback. This is slow and could cause very heavy load on ZK, especially when the number of segments are high. We should cache the intervals and only read the ZK metadata for new segments. `refreshSegment()` is used for refreshing the cached interval for existing segments.